// ldapts/filters/EqualityFilter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EqualityFilter = void 0;
const SearchFilter_1 = require("../SearchFilter");
const Filter_1 = require("./Filter");
class EqualityFilter extends Filter_1.Filter {
    constructor(options = {}) {
        super();
        this.type = SearchFilter_1.SearchFilter.equalityMatch;
        this.attribute = options.attribute || '';
        this.value = options.value || '';
    }
    parseFilter(reader) {
        this.attribute = (reader.readString() || '').toLowerCase();
        this.value = reader.readString();
        if (this.attribute === 'objectclass') {
            this.value = this.value.toLowerCase();
        }
    }
    writeFilter(writer) {
        writer.writeString(this.attribute);
        writer.writeString(this.value);
    }
    matches(objectToCheck = {}, strictAttributeCase) {
        const objectToCheckValue = this.getObjectValue(objectToCheck, this.attribute, strictAttributeCase);
        if (typeof objectToCheckValue !== 'undefined') {
            if (strictAttributeCase) {
                return this.value === objectToCheckValue;
            }
            return this.value.toLowerCase() === objectToCheckValue.toLowerCase();
        }
        return false;
    }
    toString() {
        return `(${this.escape(this.attribute)}=${this.escape(this.value)})`;
    }
}
exports.EqualityFilter = EqualityFilter;
